server/src/main/java/org/apache/iotdb/db/metadata/logfile/MLogWriter.java

```diff
@@ -0,0 +1,452 @@
+    File tmpLogFile = SystemFileFactory.INSTANCE.getFile(logFile.getAbsolutePath() + ".tmp");
+    File oldLogFile = SystemFileFactory.INSTANCE.getFile(
+      schemaDir + File.separator + oldFileName);
+    File tmpOldLogFile = SystemFileFactory.INSTANCE.getFile(oldLogFile.getAbsolutePath()
+      + ".tmp");
+
+    if (oldLogFile.exists() || tmpOldLogFile.exists()) {
+
+      if (tmpOldLogFile.exists() && !oldLogFile.exists()) {
+        FileUtils.moveFile(tmpOldLogFile, oldLogFile);
+      }
+
+      try (MLogWriter mLogWriter = new MLogWriter(schemaDir, newFileName + ".tmp");
+           MLogTxtReader mLogTxtReader = new MLogTxtReader(schemaDir, oldFileName)) {
+        // upgrade from old character log file to new binary mlog
+        while (mLogTxtReader.hasNext()) {
+          String cmd = mLogTxtReader.next();
+          try {
+            mLogWriter.operation(cmd, isSnapshot);
+          } catch (MetadataException e) {
+            logger.error("failed to upgrade cmd {}.", cmd, e);
+          }
+        }
+      }
+    } else if (!logFile.exists() && !tmpLogFile.exists()) {
+      // if both .bin and .bin.tmp do not exist, nothing to do
+    } else if (!logFile.exists() && tmpLogFile.exists()) {
+      // if old .bin doesn't exist but .bin.tmp exists, rename tmp file to .bin
+      FSFactoryProducer.getFSFactory().moveFile(tmpLogFile, logFile);
+    } else if (tmpLogFile.exists()) {
+      // if both .bin and .bin.tmp exist, delete .bin.tmp
+      try {
+        Files.delete(Paths.get(tmpLogFile.toURI()));
+      } catch (IOException e) {
+        throw new IOException(String.format(DELETE_FAILED_FORMAT, tmpLogFile, e.getMessage()));
+      }
+    }
+
+    // do some clean job
+    // remove old .txt and .txt.tmp
+    if (oldLogFile.exists()) {
+      try {
+        Files.delete(Paths.get(oldLogFile.toURI()));
+      } catch (IOException e) {
+        throw new IOException(String.format(DELETE_FAILED_FORMAT, oldLogFile, e.getMessage()));
+      }
+    }
+
+    if (tmpOldLogFile.exists()) {
+      try {
+        Files.delete(Paths.get(tmpOldLogFile.toURI()));
+      } catch (IOException e) {
+        throw new IOException(String.format(DELETE_FAILED_FORMAT, tmpOldLogFile, e.getMessage()));
+      }
+    }
+
+    // rename .bin.tmp to .bin
+    FSFactoryProducer.getFSFactory().moveFile(tmpLogFile, logFile);
+  }
+
+  public static void upgradeMLog() throws IOException {
+    String schemaDir = IoTDBDescriptor.getInstance().getConfig().getSchemaDir();
+    upgradeTxtToBin(schemaDir, MetadataConstant.METADATA_TXT_LOG, MetadataConstant.METADATA_LOG, false);
+    upgradeTxtToBin(schemaDir, MetadataConstant.MTREE_TXT_SNAPSHOT, MetadataConstant.MTREE_SNAPSHOT, true);
+  }
+
+  public void clear() throws IOException {
+    sync();
+    logWriter.close();
+    mlogBuffer.clear();
+    if (logFile != null && logFile.exists()) {
+      Files.delete(logFile.toPath());
+    }
+    logNum = 0;
+    logWriter = new LogWriter(logFile, 0L);
```

Comment:
The forcePeriodInMs is differenet between clear() and constructor.

Reply:
yes,  it's a mistake.